conmon-rs/server/src/rpc.rs

```diff
@@ -37,33 +38,44 @@ impl conmon::Server for Server {
             None
         };
 
-        let args = pry!(self.generate_runtime_args(&params, &maybe_console));
+        let pidfile = pry!(pidfile_from_params(&params));
 
-        let mut child = pry!(Command::new(self.config().runtime()).args(args).spawn());
+        let args = pry!(self
+            .generate_runtime_args(&params, &maybe_console, &pidfile)
+            .map_err(|e| Error::failed(format!("{:#}", e))));
 
-        let pid = pry!(child
-            .id()
-            .ok_or_else(|| Error::failed("Child of container creation had none id".into())));
+        let mut child = pry!(std::process::Command::new(self.config().runtime())
+            .args(args)
+            .spawn());
 
         let id = pry!(req.get_id()).to_string();
-        tokio::spawn(async move {
-            match child.wait().await {
-                Ok(status) => {
-                    debug!("Status for container ID {} is {}", id, status);
-                    if let Some(console) = maybe_console {
-                        console
-                            .wait_connected()
-                            .context("wait for console socket connection")?;
-                    }
-                    Ok(())
-                }
-                Err(e) => {
-                    error!("Failed to spawn runtime process {}", e);
-                    bail!(e)
-                }
-            }
-        });
+        let status = pry!(child.wait());
+        debug!("Status for container ID {} is {}", id, status);
+        if let Some(console) = maybe_console {
+            pry!(console
+                .wait_connected()
+                .context("wait for console socket connection")
+                .map_err(|e| Error::failed(format!("{:#}", e))));
+        }
+
+        let pid = pry!(pry!(fs::read_to_string(pidfile))
+            .parse::<u32>()
+            .map_err(|e| Error::failed(format!("{:#}", e))));
+
         results.get().init_response().set_container_pid(pid);
         Promise::ok(())
     }
 }
+
+fn pidfile_from_params(params: &conmon::CreateContainerParams) -> capnp::Result<String> {
+    let mut pidfile_pathbuf = PathBuf::from(params.get()?.get_request()?.get_bundle_path()?);
+    pidfile_pathbuf.push("pidfile");
+
+    let pidfile_str = pidfile_pathbuf.into_os_string();
```

Comment:
nit: this line and the next line can be chained all the way through.

Reply:
dropped as part of fix below